### ISSUE TITLE
bug: multiple, unrelated CTEs found during union of joined

### ISSUE THREAD

**Issue by NickCrews** [ibis-project/ibis#8288]
### What happened?

I'm not exactly sure what is the important part that actually triggers the error. I can try to simplify further if needed, but this isn't too bad.

```python
import ibis
from ibis import _
from ibis.expr.types import Table

ibis.options.interactive = True


def join(
    left: Table,
    right: Table,
    condition,
) -> Table:
    """Join two tables, afterwards adding a "_l" or "_r" suffix to all columns.

    Unlike ibis, which only adds suffixes to columns that are duplicated,
    this *always* adds suffixes to all columns.
    """
    if left is right:
        right = right.view()
    lc = set(left.columns)
    rc = set(right.columns)
    just_left = lc - rc
    just_right = rc - lc
    raw = ibis.join(left, right, condition, lname="{name}_l", rname="{name}_r")
    left_renaming = {c + "_l": c for c in just_left}
    right_renaming = {c + "_r": c for c in just_right}
    renaming = {**left_renaming, **right_renaming}
    result = raw.rename(renaming)

    # If the condition is an equality condition, like `left.name == right.name`,
    # then since we are doing an inner join ibis doesn't add suffixes to these
    # columns. So we need duplicate these columns and add suffixes.
    un_suffixed = [
        c for c in result.columns if not c.endswith("_l") and not c.endswith("_r")
    ]
    suffix_map = {c + "_l": _[c] for c in un_suffixed} | {
        c + "_r": _[c] for c in un_suffixed
    }
    result = result.mutate(**suffix_map).drop(*un_suffixed)
    return result


t = ibis.examples.penguins.fetch().select("species", "island", id=ibis.row_number())
j1 = join(t, t, "species")
j1 = j1[sorted(j1.columns)]
j2 = join(t, t, "island")
j2 = j2[sorted(j2.columns)]
j1.head(), j2.head()
# (┏━━━━━━━┳━━━━━━━┳━━━━━━━━━━━┳━━━━━━━━━━┳━━━━━━━━━━━┳━━━━━━━━━━━┓
#  ┃ id_l  ┃ id_r  ┃ island_l  ┃ island_r ┃ species_l ┃ species_r ┃
#  ┡━━━━━━━╇━━━━━━━╇━━━━━━━━━━━╇━━━━━━━━━━╇━━━━━━━━━━━╇━━━━━━━━━━━┩
#  │ int64 │ int64 │ string    │ string   │ string    │ string    │
#  ├───────┼───────┼───────────┼──────────┼───────────┼───────────┤
#  │     0 │   151 │ Torgersen │ Dream    │ Adelie    │ Adelie    │
#  │     1 │   151 │ Torgersen │ Dream    │ Adelie    │ Adelie    │
#  │     2 │   151 │ Torgersen │ Dream    │ Adelie    │ Adelie    │
#  │     3 │   151 │ Torgersen │ Dream    │ Adelie    │ Adelie    │
#  │     4 │   151 │ Torgersen │ Dream    │ Adelie    │ Adelie    │
#  └───────┴───────┴───────────┴──────────┴───────────┴───────────┘,
#  ┏━━━━━━━┳━━━━━━━┳━━━━━━━━━━━┳━━━━━━━━━━━┳━━━━━━━━━━━┳━━━━━━━━━━━┓
#  ┃ id_l  ┃ id_r  ┃ island_l  ┃ island_r  ┃ species_l ┃ species_r ┃
#  ┡━━━━━━━╇━━━━━━━╇━━━━━━━━━━━╇━━━━━━━━━━━╇━━━━━━━━━━━╇━━━━━━━━━━━┩
#  │ int64 │ int64 │ string    │ string    │ string    │ string    │
#  ├───────┼───────┼───────────┼───────────┼───────────┼───────────┤
#  │     0 │   131 │ Torgersen │ Torgersen │ Adelie    │ Adelie    │
#  │     1 │   131 │ Torgersen │ Torgersen │ Adelie    │ Adelie    │
#  │     2 │   131 │ Torgersen │ Torgersen │ Adelie    │ Adelie    │
#  │     3 │   131 │ Torgersen │ Torgersen │ Adelie    │ Adelie    │
#  │     4 │   131 │ Torgersen │ Torgersen │ Adelie    │ Adelie    │
#  └───────┴───────┴───────────┴───────────┴───────────┴───────────┘)
ibis.union(j1, j2)
```

<details>

```python-traceback
---------------------------------------------------------------------------
CompileError                              Traceback (most recent call last)
File ~/code/mismo/.venv/lib/python3.11/site-packages/IPython/core/formatters.py:708, in PlainTextFormatter.__call__(self, obj)
    701 stream = StringIO()
    702 printer = pretty.RepresentationPrinter(stream, self.verbose,
    703     self.max_width, self.newline,
    704     max_seq_length=self.max_seq_length,
    705     singleton_pprinters=self.singleton_printers,
    706     type_pprinters=self.type_printers,
    707     deferred_pprinters=self.deferred_printers)
--> 708 printer.pretty(obj)
    709 printer.flush()
    710 return stream.getvalue()

File ~/code/mismo/.venv/lib/python3.11/site-packages/IPython/lib/pretty.py:410, in RepresentationPrinter.pretty(self, obj)
    407                         return meth(obj, self, cycle)
    408                 if cls is not object \
    409                         and callable(cls.__dict__.get('__repr__')):
--> 410                     return _repr_pprint(obj, self, cycle)
    412     return _default_pprint(obj, self, cycle)
    413 finally:

File ~/code/mismo/.venv/lib/python3.11/site-packages/IPython/lib/pretty.py:778, in _repr_pprint(obj, p, cycle)
    776 """A pprint that just redirects to the normal repr function."""
    777 # Find newlines and replace them with p.break_()
--> 778 output = repr(obj)
    779 lines = output.splitlines()
    780 with p.group():

File ~/code/mismo/.venv/lib/python3.11/site-packages/ibis/expr/types/core.py:84, in Expr.__repr__(self)
     82 with simple_console.capture() as capture:
     83     try:
---> 84         simple_console.print(self)
     85     except TranslationError as e:
     86         lines = [
     87             "Translation to backend failed",
     88             f"Error message: {e.args[0]}",
     89             "Expression repr follows:",
     90             self._repr(),
     91         ]

File ~/code/mismo/.venv/lib/python3.11/site-packages/rich/console.py:1700, in Console.print(self, sep, end, style, justify, overflow, no_wrap, emoji, markup, highlight, width, height, crop, soft_wrap, new_line_start, *objects)
   1698 if style is None:
   1699     for renderable in renderables:
-> 1700         extend(render(renderable, render_options))
   1701 else:
   1702     for renderable in renderables:

File ~/code/mismo/.venv/lib/python3.11/site-packages/rich/console.py:1312, in Console.render(self, renderable, options)
   1310 renderable = rich_cast(renderable)
   1311 if hasattr(renderable, "__rich_console__") and not isclass(renderable):
-> 1312     render_iterable = renderable.__rich_console__(self, _options)  # type: ignore[union-attr]
   1313 elif isinstance(renderable, str):
   1314     text_renderable = self.render_str(
   1315         renderable, highlight=_options.highlight, markup=_options.markup
   1316     )

File ~/code/mismo/.venv/lib/python3.11/site-packages/ibis/expr/types/core.py:56, in Expr.__rich_console__(self, console, options)
     53     from rich.text import Text
     55     return console.render(Text(self._repr()), options=options)
---> 56 return self.__interactive_rich_console__(console, options)

File ~/code/mismo/.venv/lib/python3.11/site-packages/ibis/expr/types/relations.py:387, in Table.__interactive_rich_console__(self, console, options)
    371 except Exception as e:
    372     # In IPython exceptions inside of _repr_mimebundle_ are swallowed to
    373     # allow calling several display functions and choosing to display
   (...)
    384     #
    385     # This restriction is only present in IPython, not in other REPLs.
    386     console.print_exception()
--> 387     raise e
    388 return console.render(table, options=options)

File ~/code/mismo/.venv/lib/python3.11/site-packages/ibis/expr/types/relations.py:370, in Table.__interactive_rich_console__(self, console, options)
    367     width = options.max_width
    369 try:
--> 370     table = to_rich_table(self, width)
    371 except Exception as e:
    372     # In IPython exceptions inside of _repr_mimebundle_ are swallowed to
    373     # allow calling several display functions and choosing to display
   (...)
    384     #
    385     # This restriction is only present in IPython, not in other REPLs.
    386     console.print_exception()

File ~/code/mismo/.venv/lib/python3.11/site-packages/ibis/expr/types/pretty.py:265, in to_rich_table(table, console_width)
    263 # Compute the data and return a pandas dataframe
    264 nrows = ibis.options.repr.interactive.max_rows
--> 265 result = table.limit(nrows + 1).to_pyarrow()
    267 # Now format the columns in order, stopping if the console width would
    268 # be exceeded.
    269 col_info = []

File ~/code/mismo/.venv/lib/python3.11/site-packages/ibis/expr/types/core.py:423, in Expr.to_pyarrow(self, params, limit, **kwargs)
    395 @experimental
    396 def to_pyarrow(
    397     self,
   (...)
    401     **kwargs: Any,
    402 ) -> pa.Table:
    403     """Execute expression and return results in as a pyarrow table.
    404 
    405     This method is eager and will execute the associated expression
   (...)
    421         A pyarrow table holding the results of the executed expression.
    422     """
--> 423     return self._find_backend(use_default=True).to_pyarrow(
    424         self, params=params, limit=limit, **kwargs
    425     )

File ~/code/mismo/.venv/lib/python3.11/site-packages/ibis/backends/duckdb/__init__.py:1040, in Backend.to_pyarrow(self, expr, params, limit, **_)
   1033 query_ast = self.compiler.to_ast_ensure_limit(expr, limit, params=params)
   1035 # We use `.sql` instead of `.execute` below for performance - in
   1036 # certain cases duckdb query -> arrow table can be significantly faster
   1037 # in this configuration. Currently `.sql` doesn't support parametrized
   1038 # queries, so we need to compile with literal_binds for now.
   1039 sql = str(
-> 1040     query_ast.compile().compile(
   1041         dialect=self.con.dialect, compile_kwargs={"literal_binds": True}
   1042     )
   1043 )
   1045 with self.begin() as con:
   1046     table = con.connection.sql(sql).to_arrow_table()

File ~/code/mismo/.venv/lib/python3.11/site-packages/sqlalchemy/sql/elements.py:309, in CompilerElement.compile(self, bind, dialect, **kw)
    304         url = util.preloaded.engine_url
    305         dialect = url.URL.create(
    306             self.stringify_dialect
    307         ).get_dialect()()
--> 309 return self._compiler(dialect, **kw)

File ~/code/mismo/.venv/lib/python3.11/site-packages/sqlalchemy/sql/elements.py:317, in CompilerElement._compiler(self, dialect, **kw)
    315 if TYPE_CHECKING:
    316     assert isinstance(self, ClauseElement)
--> 317 return dialect.statement_compiler(dialect, self, **kw)

File ~/code/mismo/.venv/lib/python3.11/site-packages/sqlalchemy/sql/compiler.py:1427, in SQLCompiler.__init__(self, dialect, statement, cache_key, column_keys, for_executemany, linting, _supporting_against, **kwargs)
   1424 self.truncated_names: Dict[Tuple[str, str], str] = {}
   1425 self._truncated_counters: Dict[str, int] = {}
-> 1427 Compiled.__init__(self, dialect, statement, **kwargs)
   1429 if self.isinsert or self.isupdate or self.isdelete:
   1430     if TYPE_CHECKING:

File ~/code/mismo/.venv/lib/python3.11/site-packages/sqlalchemy/sql/compiler.py:867, in Compiled.__init__(self, dialect, statement, schema_translate_map, render_schema_translate, compile_kwargs)
    865         assert isinstance(statement, Executable)
    866     self.execution_options = statement._execution_options
--> 867 self.string = self.process(self.statement, **compile_kwargs)
    869 if render_schema_translate:
    870     self.string = self.preparer._render_schema_translates(
    871         self.string, schema_translate_map
    872     )

File ~/code/mismo/.venv/lib/python3.11/site-packages/sqlalchemy/sql/compiler.py:912, in Compiled.process(self, obj, **kwargs)
    911 def process(self, obj: Visitable, **kwargs: Any) -> str:
--> 912     return obj._compiler_dispatch(self, **kwargs)

File ~/code/mismo/.venv/lib/python3.11/site-packages/sqlalchemy/sql/visitors.py:143, in Visitable._generate_compiler_dispatch.<locals>._compiler_dispatch(self, visitor, **kw)
    141     return visitor.visit_unsupported_compilation(self, err, **kw)  # type: ignore  # noqa: E501
    142 else:
--> 143     return meth(self, **kw)

File ~/code/mismo/.venv/lib/python3.11/site-packages/sqlalchemy/sql/compiler.py:4754, in SQLCompiler.visit_select(self, select_stmt, asfrom, insert_into, fromhints, compound_index, select_wraps_for, lateral, from_linter, **kwargs)
   4722     translate = dict(
   4723         zip(
   4724             [
   (...)
   4744         )
   4745     )
   4747     self._result_columns = [
   4748         ResultColumnsEntry(
   4749             key, name, tuple(translate.get(o, o) for o in obj), type_
   4750         )
   4751         for key, name, obj, type_ in self._result_columns
   4752     ]
-> 4754 text = self._compose_select_body(
   4755     text,
   4756     select_stmt,
   4757     compile_state,
   4758     inner_columns,
   4759     froms,
   4760     byfrom,
   4761     toplevel,
   4762     kwargs,
   4763 )
   4765 if select_stmt._statement_hints:
   4766     per_dialect = [
   4767         ht
   4768         for (dialect_name, ht) in select_stmt._statement_hints
   4769         if dialect_name in ("*", self.dialect.name)
   4770     ]

File ~/code/mismo/.venv/lib/python3.11/site-packages/sqlalchemy/sql/compiler.py:4902, in SQLCompiler._compose_select_body(self, text, select, compile_state, inner_columns, froms, byfrom, toplevel, kwargs)
   4888         text += ", ".join(
   4889             [
   4890                 f._compiler_dispatch(
   (...)
   4898             ]
   4899         )
   4900     else:
   4901         text += ", ".join(
-> 4902             [
   4903                 f._compiler_dispatch(
   4904                     self,
   4905                     asfrom=True,
   4906                     from_linter=from_linter,
   4907                     **kwargs,
   4908                 )
   4909                 for f in froms
   4910             ]
   4911         )
   4912 else:
   4913     text += self.default_from()

File ~/code/mismo/.venv/lib/python3.11/site-packages/sqlalchemy/sql/compiler.py:4903, in <listcomp>(.0)
   4888         text += ", ".join(
   4889             [
   4890                 f._compiler_dispatch(
   (...)
   4898             ]
   4899         )
   4900     else:
   4901         text += ", ".join(
   4902             [
-> 4903                 f._compiler_dispatch(
   4904                     self,
   4905                     asfrom=True,
   4906                     from_linter=from_linter,
   4907                     **kwargs,
   4908                 )
   4909                 for f in froms
   4910             ]
   4911         )
   4912 else:
   4913     text += self.default_from()

File ~/code/mismo/.venv/lib/python3.11/site-packages/sqlalchemy/sql/visitors.py:143, in Visitable._generate_compiler_dispatch.<locals>._compiler_dispatch(self, visitor, **kw)
    141     return visitor.visit_unsupported_compilation(self, err, **kw)  # type: ignore  # noqa: E501
    142 else:
--> 143     return meth(self, **kw)

File ~/code/mismo/.venv/lib/python3.11/site-packages/sqlalchemy/sql/compiler.py:4240, in SQLCompiler.visit_subquery(self, subquery, **kw)
   4238 def visit_subquery(self, subquery, **kw):
   4239     kw["subquery"] = True
-> 4240     return self.visit_alias(subquery, **kw)

File ~/code/mismo/.venv/lib/python3.11/site-packages/sqlalchemy/sql/compiler.py:4199, in SQLCompiler.visit_alias(self, alias, asfrom, ashint, iscrud, fromhints, subquery, lateral, enclosing_alias, from_linter, **kwargs)
   4196 if from_linter:
   4197     from_linter.froms[alias._de_clone()] = alias_name
-> 4199 inner = alias.element._compiler_dispatch(
   4200     self, asfrom=True, lateral=lateral, **kwargs
   4201 )
   4202 if subquery:
   4203     inner = "(%s)" % (inner,)

File ~/code/mismo/.venv/lib/python3.11/site-packages/sqlalchemy/sql/visitors.py:143, in Visitable._generate_compiler_dispatch.<locals>._compiler_dispatch(self, visitor, **kw)
    141     return visitor.visit_unsupported_compilation(self, err, **kw)  # type: ignore  # noqa: E501
    142 else:
--> 143     return meth(self, **kw)

File ~/code/mismo/.venv/lib/python3.11/site-packages/sqlalchemy/sql/compiler.py:2968, in SQLCompiler.visit_compound_select(self, cs, asfrom, compound_index, **kwargs)
   2964     self._dispatch_independent_ctes(compound_stmt, kwargs)
   2966 keyword = self.compound_keywords[cs.keyword]
-> 2968 text = (" " + keyword + " ").join(
   2969     (
   2970         c._compiler_dispatch(
   2971             self, asfrom=asfrom, compound_index=i, **kwargs
   2972         )
   2973         for i, c in enumerate(cs.selects)
   2974     )
   2975 )
   2977 kwargs["include_table"] = False
   2978 text += self.group_by_clause(cs, **dict(asfrom=asfrom, **kwargs))

File ~/code/mismo/.venv/lib/python3.11/site-packages/sqlalchemy/sql/compiler.py:2970, in <genexpr>(.0)
   2964     self._dispatch_independent_ctes(compound_stmt, kwargs)
   2966 keyword = self.compound_keywords[cs.keyword]
   2968 text = (" " + keyword + " ").join(
   2969     (
-> 2970         c._compiler_dispatch(
   2971             self, asfrom=asfrom, compound_index=i, **kwargs
   2972         )
   2973         for i, c in enumerate(cs.selects)
   2974     )
   2975 )
   2977 kwargs["include_table"] = False
   2978 text += self.group_by_clause(cs, **dict(asfrom=asfrom, **kwargs))

File ~/code/mismo/.venv/lib/python3.11/site-packages/sqlalchemy/sql/visitors.py:143, in Visitable._generate_compiler_dispatch.<locals>._compiler_dispatch(self, visitor, **kw)
    141     return visitor.visit_unsupported_compilation(self, err, **kw)  # type: ignore  # noqa: E501
    142 else:
--> 143     return meth(self, **kw)

File ~/code/mismo/.venv/lib/python3.11/site-packages/sqlalchemy/sql/compiler.py:4754, in SQLCompiler.visit_select(self, select_stmt, asfrom, insert_into, fromhints, compound_index, select_wraps_for, lateral, from_linter, **kwargs)
   4722     translate = dict(
   4723         zip(
   4724             [
   (...)
   4744         )
   4745     )
   4747     self._result_columns = [
   4748         ResultColumnsEntry(
   4749             key, name, tuple(translate.get(o, o) for o in obj), type_
   4750         )
   4751         for key, name, obj, type_ in self._result_columns
   4752     ]
-> 4754 text = self._compose_select_body(
   4755     text,
   4756     select_stmt,
   4757     compile_state,
   4758     inner_columns,
   4759     froms,
   4760     byfrom,
   4761     toplevel,
   4762     kwargs,
   4763 )
   4765 if select_stmt._statement_hints:
   4766     per_dialect = [
   4767         ht
   4768         for (dialect_name, ht) in select_stmt._statement_hints
   4769         if dialect_name in ("*", self.dialect.name)
   4770     ]

File ~/code/mismo/.venv/lib/python3.11/site-packages/sqlalchemy/sql/compiler.py:4902, in SQLCompiler._compose_select_body(self, text, select, compile_state, inner_columns, froms, byfrom, toplevel, kwargs)
   4888         text += ", ".join(
   4889             [
   4890                 f._compiler_dispatch(
   (...)
   4898             ]
   4899         )
   4900     else:
   4901         text += ", ".join(
-> 4902             [
   4903                 f._compiler_dispatch(
   4904                     self,
   4905                     asfrom=True,
   4906                     from_linter=from_linter,
   4907                     **kwargs,
   4908                 )
   4909                 for f in froms
   4910             ]
   4911         )
   4912 else:
   4913     text += self.default_from()

File ~/code/mismo/.venv/lib/python3.11/site-packages/sqlalchemy/sql/compiler.py:4903, in <listcomp>(.0)
   4888         text += ", ".join(
   4889             [
   4890                 f._compiler_dispatch(
   (...)
   4898             ]
   4899         )
   4900     else:
   4901         text += ", ".join(
   4902             [
-> 4903                 f._compiler_dispatch(
   4904                     self,
   4905                     asfrom=True,
   4906                     from_linter=from_linter,
   4907                     **kwargs,
   4908                 )
   4909                 for f in froms
   4910             ]
   4911         )
   4912 else:
   4913     text += self.default_from()

File ~/code/mismo/.venv/lib/python3.11/site-packages/sqlalchemy/sql/visitors.py:143, in Visitable._generate_compiler_dispatch.<locals>._compiler_dispatch(self, visitor, **kw)
    141     return visitor.visit_unsupported_compilation(self, err, **kw)  # type: ignore  # noqa: E501
    142 else:
--> 143     return meth(self, **kw)

File ~/code/mismo/.venv/lib/python3.11/site-packages/sqlalchemy/sql/compiler.py:4098, in SQLCompiler.visit_cte(self, cte, asfrom, ashint, fromhints, visiting_cte, from_linter, cte_opts, **kwargs)
   4094 else:
   4095     prefixes = self._generate_prefixes(
   4096         cte, cte._prefixes, **kwargs
   4097     )
-> 4098     inner = cte.element._compiler_dispatch(
   4099         self, asfrom=True, **kwargs
   4100     )
   4102     text += " AS %s\n(%s)" % (prefixes, inner)
   4104 if cte._suffixes:

    [... skipping similar frames: Visitable._generate_compiler_dispatch.<locals>._compiler_dispatch at line 143 (1 times)]

File ~/code/mismo/.venv/lib/python3.11/site-packages/sqlalchemy/sql/compiler.py:4754, in SQLCompiler.visit_select(self, select_stmt, asfrom, insert_into, fromhints, compound_index, select_wraps_for, lateral, from_linter, **kwargs)
   4722     translate = dict(
   4723         zip(
   4724             [
   (...)
   4744         )
   4745     )
   4747     self._result_columns = [
   4748         ResultColumnsEntry(
   4749             key, name, tuple(translate.get(o, o) for o in obj), type_
   4750         )
   4751         for key, name, obj, type_ in self._result_columns
   4752     ]
-> 4754 text = self._compose_select_body(
   4755     text,
   4756     select_stmt,
   4757     compile_state,
   4758     inner_columns,
   4759     froms,
   4760     byfrom,
   4761     toplevel,
   4762     kwargs,
   4763 )
   4765 if select_stmt._statement_hints:
   4766     per_dialect = [
   4767         ht
   4768         for (dialect_name, ht) in select_stmt._statement_hints
   4769         if dialect_name in ("*", self.dialect.name)
   4770     ]

File ~/code/mismo/.venv/lib/python3.11/site-packages/sqlalchemy/sql/compiler.py:4902, in SQLCompiler._compose_select_body(self, text, select, compile_state, inner_columns, froms, byfrom, toplevel, kwargs)
   4888         text += ", ".join(
   4889             [
   4890                 f._compiler_dispatch(
   (...)
   4898             ]
   4899         )
   4900     else:
   4901         text += ", ".join(
-> 4902             [
   4903                 f._compiler_dispatch(
   4904                     self,
   4905                     asfrom=True,
   4906                     from_linter=from_linter,
   4907                     **kwargs,
   4908                 )
   4909                 for f in froms
   4910             ]
   4911         )
   4912 else:
   4913     text += self.default_from()

File ~/code/mismo/.venv/lib/python3.11/site-packages/sqlalchemy/sql/compiler.py:4903, in <listcomp>(.0)
   4888         text += ", ".join(
   4889             [
   4890                 f._compiler_dispatch(
   (...)
   4898             ]
   4899         )
   4900     else:
   4901         text += ", ".join(
   4902             [
-> 4903                 f._compiler_dispatch(
   4904                     self,
   4905                     asfrom=True,
   4906                     from_linter=from_linter,
   4907                     **kwargs,
   4908                 )
   4909                 for f in froms
   4910             ]
   4911         )
   4912 else:
   4913     text += self.default_from()

    [... skipping similar frames: Visitable._generate_compiler_dispatch.<locals>._compiler_dispatch at line 143 (1 times)]

File ~/code/mismo/.venv/lib/python3.11/site-packages/sqlalchemy/sql/compiler.py:4098, in SQLCompiler.visit_cte(self, cte, asfrom, ashint, fromhints, visiting_cte, from_linter, cte_opts, **kwargs)
   4094 else:
   4095     prefixes = self._generate_prefixes(
   4096         cte, cte._prefixes, **kwargs
   4097     )
-> 4098     inner = cte.element._compiler_dispatch(
   4099         self, asfrom=True, **kwargs
   4100     )
   4102     text += " AS %s\n(%s)" % (prefixes, inner)
   4104 if cte._suffixes:

    [... skipping similar frames: Visitable._generate_compiler_dispatch.<locals>._compiler_dispatch at line 143 (6 times), <listcomp> at line 4903 (3 times), SQLCompiler._compose_select_body at line 4902 (3 times), SQLCompiler.visit_select at line 4754 (3 times), SQLCompiler.visit_cte at line 4098 (2 times)]

File ~/code/mismo/.venv/lib/python3.11/site-packages/sqlalchemy/sql/compiler.py:4098, in SQLCompiler.visit_cte(self, cte, asfrom, ashint, fromhints, visiting_cte, from_linter, cte_opts, **kwargs)
   4094 else:
   4095     prefixes = self._generate_prefixes(
   4096         cte, cte._prefixes, **kwargs
   4097     )
-> 4098     inner = cte.element._compiler_dispatch(
   4099         self, asfrom=True, **kwargs
   4100     )
   4102     text += " AS %s\n(%s)" % (prefixes, inner)
   4104 if cte._suffixes:

    [... skipping similar frames: Visitable._generate_compiler_dispatch.<locals>._compiler_dispatch at line 143 (1 times)]

File ~/code/mismo/.venv/lib/python3.11/site-packages/sqlalchemy/sql/compiler.py:4754, in SQLCompiler.visit_select(self, select_stmt, asfrom, insert_into, fromhints, compound_index, select_wraps_for, lateral, from_linter, **kwargs)
   4722     translate = dict(
   4723         zip(
   4724             [
   (...)
   4744         )
   4745     )
   4747     self._result_columns = [
   4748         ResultColumnsEntry(
   4749             key, name, tuple(translate.get(o, o) for o in obj), type_
   4750         )
   4751         for key, name, obj, type_ in self._result_columns
   4752     ]
-> 4754 text = self._compose_select_body(
   4755     text,
   4756     select_stmt,
   4757     compile_state,
   4758     inner_columns,
   4759     froms,
   4760     byfrom,
   4761     toplevel,
   4762     kwargs,
   4763 )
   4765 if select_stmt._statement_hints:
   4766     per_dialect = [
   4767         ht
   4768         for (dialect_name, ht) in select_stmt._statement_hints
   4769         if dialect_name in ("*", self.dialect.name)
   4770     ]

File ~/code/mismo/.venv/lib/python3.11/site-packages/sqlalchemy/sql/compiler.py:4902, in SQLCompiler._compose_select_body(self, text, select, compile_state, inner_columns, froms, byfrom, toplevel, kwargs)
   4888         text += ", ".join(
   4889             [
   4890                 f._compiler_dispatch(
   (...)
   4898             ]
   4899         )
   4900     else:
   4901         text += ", ".join(
-> 4902             [
   4903                 f._compiler_dispatch(
   4904                     self,
   4905                     asfrom=True,
   4906                     from_linter=from_linter,
   4907                     **kwargs,
   4908                 )
   4909                 for f in froms
   4910             ]
   4911         )
   4912 else:
   4913     text += self.default_from()

File ~/code/mismo/.venv/lib/python3.11/site-packages/sqlalchemy/sql/compiler.py:4903, in <listcomp>(.0)
   4888         text += ", ".join(
   4889             [
   4890                 f._compiler_dispatch(
   (...)
   4898             ]
   4899         )
   4900     else:
   4901         text += ", ".join(
   4902             [
-> 4903                 f._compiler_dispatch(
   4904                     self,
   4905                     asfrom=True,
   4906                     from_linter=from_linter,
   4907                     **kwargs,
   4908                 )
   4909                 for f in froms
   4910             ]
   4911         )
   4912 else:
   4913     text += self.default_from()

File ~/code/mismo/.venv/lib/python3.11/site-packages/sqlalchemy/sql/visitors.py:143, in Visitable._generate_compiler_dispatch.<locals>._compiler_dispatch(self, visitor, **kw)
    141     return visitor.visit_unsupported_compilation(self, err, **kw)  # type: ignore  # noqa: E501
    142 else:
--> 143     return meth(self, **kw)

File ~/code/mismo/.venv/lib/python3.11/site-packages/sqlalchemy/sql/compiler.py:5209, in SQLCompiler.visit_join(self, join, asfrom, from_linter, **kwargs)
   5202 else:
   5203     join_type = " JOIN "
   5204 return (
   5205     join.left._compiler_dispatch(
   5206         self, asfrom=True, from_linter=from_linter, **kwargs
   5207     )
   5208     + join_type
-> 5209     + join.right._compiler_dispatch(
   5210         self, asfrom=True, from_linter=from_linter, **kwargs
   5211     )
   5212     + " ON "
   5213     # TODO: likely need asfrom=True here?
   5214     + join.onclause._compiler_dispatch(
   5215         self, from_linter=from_linter, **kwargs
   5216     )
   5217 )

File ~/code/mismo/.venv/lib/python3.11/site-packages/sqlalchemy/sql/visitors.py:143, in Visitable._generate_compiler_dispatch.<locals>._compiler_dispatch(self, visitor, **kw)
    141     return visitor.visit_unsupported_compilation(self, err, **kw)  # type: ignore  # noqa: E501
    142 else:
--> 143     return meth(self, **kw)

File ~/code/mismo/.venv/lib/python3.11/site-packages/sqlalchemy/sql/compiler.py:4022, in SQLCompiler.visit_cte(self, cte, asfrom, ashint, fromhints, visiting_cte, from_linter, cte_opts, **kwargs)
   4020             is_new_cte = False
   4021         else:
-> 4022             raise exc.CompileError(
   4023                 "Multiple, unrelated CTEs found with "
   4024                 "the same name: %r" % cte_name
   4025             )
   4027 if not asfrom and not is_new_cte:
   4028     return None

CompileError: Multiple, unrelated CTEs found with the same name: 't10'
```

<details>

### What version of ibis are you using?

8.0.0

### What backend(s) are you using, if any?

duckdb

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

**Comment by cpcloud** [comment]
@NickCrews Can you try against `the-epic-split`?

**Comment by NickCrews** [comment]
oh, nice, it runs with no error there!

**Comment by NickCrews** [comment]
I just tried pinning my project to use a recent commit from `the-epic-split` so I could avoid this bug, but that leads to a an error right away. Is that branch not stable? Do I need to wait for it to get merged into main before it should be considered usable?

**Comment by cpcloud** [comment]
If you feel comfortable reporting that error we can hopefully fix it ASAP.

**Comment by cpcloud** [comment]
@NickCrews Yes, `the-epic-split` is not stable. Once it is merged into `main` there are likely still plenty of follow up tasks before we release 9.0.

**Comment by cpcloud** [comment]
Closing this as working on `the-epic-split`.
